File: app/src/main/java/com/bitflaker/lucidsourcekit/setup/SetupViewer.java
package com.bitflaker.lucidsourcekit.setup;

import android.content.Intent;
import android.os.Bundle;
import android.util.Base64;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.res.ResourcesCompat;
import androidx.viewpager2.widget.ViewPager2;

import com.bitflaker.lucidsourcekit.R;
import com.bitflaker.lucidsourcekit.data.datastore.DataStoreKeys;
import com.bitflaker.lucidsourcekit.data.datastore.DataStoreManager;
import com.bitflaker.lucidsourcekit.data.enums.AuthTypes;
import com.bitflaker.lucidsourcekit.databinding.ActivitySetupViewerBinding;
import com.bitflaker.lucidsourcekit.utils.Crypt;
import com.bitflaker.lucidsourcekit.utils.Tools;
import com.google.android.material.tabs.TabLayout;

import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.util.ArrayList;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

public class SetupViewer extends AppCompatActivity {
    private final ArrayList<ImageView> pageDots = new ArrayList<>();
    private AuthTypes selectedAuthType = AuthTypes.Pin;

    private final AtomicBoolean consented = new AtomicBoolean(false);
    private final int pageCount = 5;
    private final String pageLang = "Language";
    private final String pageConsent = "Consent";
    private final String pageOpenSource = "OpenSource";
    private final String pagePrivacy = "Privacy";
    private final String pageAuthData = "AuthData";

    private ViewPagerAdapter vpAdapter;
    private final SetupLanguageView setLang = new SetupLanguageView();
    private final SetupConsentView setConsent = new SetupConsentView();
    private final SetupOpenSourceView setOpenSource = new SetupOpenSourceView();
    private final SetupPrivacyView setPrivacy = new SetupPrivacyView();
    private final SetupAuthDataView setAuthData = new SetupAuthDataView();

    private final AtomicReference<TabLayout.Tab> tabOpenSource = new AtomicReference<>();
    private final AtomicReference<TabLayout.Tab> tabPrivacy = new AtomicReference<>();
    private final AtomicReference<TabLayout.Tab> tabAuthData = new AtomicReference<>();

    private ActivitySetupViewerBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySetupViewerBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        Tools.makeStatusBarTransparent(this);

        setPrivacy.setOnAuthTypeChangedListener(authType -> {
            selectedAuthType = authType;
            setAuthData.showCredentialsSetup(authType);
        });

        setLang.setLanguageChangedListener(this::languageChanged);
        setConsent.setOnConsentChangedListener(this::consentChanged);

        vpAdapter = new ViewPagerAdapter(getSupportFragmentManager(), getLifecycle());
        vpAdapter.addFragment(setLang, pageLang);
        vpAdapter.addFragment(setConsent, pageConsent);
        binding.viewpager.setAdapter(vpAdapter);

        binding.tablayout.addTab(binding.tablayout.newTab().setText(pageLang));
        binding.tablayout.addTab(binding.tablayout.newTab().setText(pageConsent));

        addProgressDots();

        binding.tablayout.addOnTabSelectedListener(tabSelected());
        binding.viewpager.registerOnPageChangeCallback(changeTab());
        binding.btnNext.setOnClickListener(this::progressToNextPage);
    }

    private void progressToNextPage(View view) {
        // Invalid pages
        if (binding.tablayout.getSelectedTabPosition() < 0 || binding.tablayout.getSelectedTabPosition() >= vpAdapter.getItemCount()) {
            Log.e("SetupViewer", "Invalid tab position \"" + binding.tablayout.getSelectedTabPosition() + "\" on item count \"" + vpAdapter.getItemCount() + "\"");
            return;
        }

        // Handle every page except the AuthData page
        if (!vpAdapter.isCurrentPage(pageAuthData, binding.tablayout.getSelectedTabPosition())) {
            int nextPagePosition = binding.tablayout.getSelectedTabPosition() + 1;
            if (nextPagePosition < vpAdapter.getItemCount()) {
                binding.tablayout.selectTab(binding.tablayout.getTabAt(nextPagePosition));
            }
        }
        else {
            handleAuthenticationPage();
        }
    }

    private void handleAuthenticationPage() {
        try {
            switch (selectedAuthType) {
                case Password: storePassword(); break;
                case Pin: storePIN(); break;
                case None: storeWithoutAuthentication(); break;
            }
            startGetStarted();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    private void storeWithoutAuthentication() {
        DataStoreManager dsManager = DataStoreManager.getInstance();
        dsManager.updateSetting(DataStoreKeys.AUTHENTICATION_TYPE, "none").blockingSubscribe();
    }

    private void storePIN() throws NoSuchAlgorithmException {
        if (setAuthData.getPin().isEmpty()) {
            Toast.makeText(SetupViewer.this, getResources().getString(R.string.pin_too_short), Toast.LENGTH_SHORT).show();
            return;
        }
        DataStoreManager dsManager = DataStoreManager.getInstance();
        byte[] secretKey = Crypt.generateSecretKey();
        String pinCipher = Crypt.encryptStringBlowfish(setAuthData.getPin(), secretKey);
        dsManager.updateSetting(DataStoreKeys.AUTHENTICATION_TYPE, "pin").blockingSubscribe();
        dsManager.updateSetting(DataStoreKeys.AUTHENTICATION_HASH, pinCipher).blockingSubscribe();
        dsManager.updateSetting(DataStoreKeys.AUTHENTICATION_SALT, Base64.encodeToString(secretKey, Base64.NO_WRAP)).blockingSubscribe();
        dsManager.updateSetting(DataStoreKeys.AUTHENTICATION_USE_BIOMETRICS, setPrivacy.getUseBiometrics()).blockingSubscribe();
    }

    private void storePassword() throws NoSuchAlgorithmException, InvalidKeySpecException {
        if (setAuthData.getPassword().isEmpty()) {
            Toast.makeText(SetupViewer.this, getResources().getString(R.string.password_too_short), Toast.LENGTH_SHORT).show();
            return;
        }
        DataStoreManager dsManager = DataStoreManager.getInstance();
        byte[] salt = Crypt.generateSalt();
        String pwHash = Crypt.encryptString(setAuthData.getPassword(), salt);
        dsManager.updateSetting(DataStoreKeys.AUTHENTICATION_TYPE, "password").blockingSubscribe();
        dsManager.updateSetting(DataStoreKeys.AUTHENTICATION_HASH, pwHash).blockingSubscribe();
        dsManager.updateSetting(DataStoreKeys.AUTHENTICATION_SALT, Base64.encodeToString(salt, Base64.NO_WRAP)).blockingSubscribe();
        dsManager.updateSetting(DataStoreKeys.AUTHENTICATION_USE_BIOMETRICS, setPrivacy.getUseBiometrics()).blockingSubscribe();
    }

    @NonNull
    private ViewPager2.OnPageChangeCallback changeTab() {
        return new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                binding.tablayout.selectTab(binding.tablayout.getTabAt(position));
            }
        };
    }

    @NonNull
    private TabLayout.OnTabSelectedListener tabSelected() {
        return new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                binding.viewpager.setCurrentItem(tab.getPosition());
                pageDots.get(tab.getPosition()).setBackgroundTintList(Tools.getAttrColorStateList(R.attr.colorPrimary, getTheme()));

                if (binding.tablayout.getSelectedTabPosition() + 1 == pageCount) {
                    binding.btnNext.setText(R.string.setup_next_finish);
                    binding.btnNext.setIcon(null);
                }
                else {
                    binding.btnNext.setText(R.string.setup_next);
                    binding.btnNext.setIcon(ResourcesCompat.getDrawable(getResources(), R.drawable.rounded_chevron_right_24, getTheme()));
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
                pageDots.get(tab.getPosition()).setBackgroundTintList(Tools.getAttrColorStateList(R.attr.colorSurfaceContainerHigh, getTheme()));
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) { }
        };
    }

    private void addProgressDots() {
        for (int i = 0; i < pageCount; i++) {
            ImageView dotPage = new ImageView(this);
            dotPage.setScaleType(ImageView.ScaleType.FIT_CENTER);
            dotPage.setBackgroundResource(R.drawable.page_dot);
            RelativeLayout.LayoutParams lp = new RelativeLayout.LayoutParams(Tools.dpToPx(this, 8), Tools.dpToPx(this, 8));
            lp.setMargins(Tools.dpToPx(this, 2), 0, Tools.dpToPx(this, 2), 0);
            dotPage.setBackgroundTintList(Tools.getAttrColorStateList(i == 0 ? R.attr.colorPrimary : R.attr.colorSurfaceContainerHigh, getTheme()));
            binding.pageDotContainer.addView(dotPage, lp);
            pageDots.add(dotPage);
        }
    }

    private void consentChanged(boolean currentConsent) {
        if (currentConsent) {
            vpAdapter.addFragment(setOpenSource, pageOpenSource);
            vpAdapter.addFragment(setPrivacy, pagePrivacy);
            vpAdapter.addFragment(setAuthData, pageAuthData);

            tabOpenSource.set(binding.tablayout.newTab().setText(pageOpenSource));
            tabPrivacy.set(binding.tablayout.newTab().setText(pagePrivacy));
            tabAuthData.set(binding.tablayout.newTab().setText(pageAuthData));

            binding.tablayout.addTab(tabOpenSource.get());
            binding.tablayout.addTab(tabPrivacy.get());
            binding.tablayout.addTab(tabAuthData.get());

            consented.set(true);
        }
        else {
            vpAdapter.removeFragment(setOpenSource, pageOpenSource);
            vpAdapter.removeFragment(setPrivacy, pagePrivacy);
            vpAdapter.removeFragment(setAuthData, pageAuthData);

            binding.tablayout.removeTab(tabOpenSource.get());
            binding.tablayout.removeTab(tabPrivacy.get());
            binding.tablayout.removeTab(tabAuthData.get());

            consented.set(false);
        }
    }

    private void languageChanged(String langCode) {
        binding.btnNext.setText(getResources().getString(R.string.setup_next));
        ((SetupLanguageView) vpAdapter.getFragment(pageLang)).updateLanguages();
        ((SetupConsentView) vpAdapter.getFragment(pageConsent)).updateLanguages();
        if (consented.get()) {
            ((SetupOpenSourceView) vpAdapter.getFragment(pageOpenSource)).updateLanguages();
            ((SetupPrivacyView) vpAdapter.getFragment(pagePrivacy)).updateLanguages();
            ((SetupAuthDataView) vpAdapter.getFragment(pageAuthData)).updateLanguages();
        }
    }

    public void startGetStarted() {
        DataStoreManager.getInstance().updateSetting(DataStoreKeys.APP_SETUP_FINISHED, true).blockingSubscribe();
        Intent intent = new Intent(SetupViewer.this, SetupGetStartedView.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
        finish();
    }
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/setup/SetupViewer.kt
package com.bitflaker.lucidsourcekit.setup

import android.content.Intent
import android.os.Bundle
import android.util.Base64
import android.view.View
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.lifecycleScope
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.bitflaker.lucidsourcekit.R
import com.bitflaker.lucidsourcekit.data.datastore.DataStoreKeys
import com.bitflaker.lucidsourcekit.data.datastore.updateSetting
import com.bitflaker.lucidsourcekit.data.enums.AuthTypes
import com.bitflaker.lucidsourcekit.databinding.ActivitySetupViewerBinding
import com.bitflaker.lucidsourcekit.utils.Crypt
import com.bitflaker.lucidsourcekit.utils.attrColorStateList
import com.bitflaker.lucidsourcekit.utils.dpToPx
import com.bitflaker.lucidsourcekit.utils.resolveDrawable
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.OnTabSelectedListener
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

private const val PAGE_LABEL_LANGUAGE = "Language"
private const val PAGE_LABEL_CONSENT = "Consent"
private const val PAGE_LABEL_OPEN_SOURCE = "OpenSource"
private const val PAGE_LABEL_PRIVACY = "Privacy"
private const val PAGE_LABEL_AUTHENTICATION = "AuthData"

class SetupViewer : AppCompatActivity() {
    private val pageDots = ArrayList<ImageView>()
    private val consented = AtomicBoolean(false)
    private var selectedAuthType = AuthTypes.Pin

    private lateinit var vpAdapter: ViewPagerAdapter
    private val pageLanguage = SetupLanguageView()
    private val pageConsent = SetupConsentView()
    private val pageOpenSource = SetupOpenSourceView()
    private val pagePrivacy = SetupPrivacyView()
    private val pageAuthentication = SetupAuthDataView()
    private val pageCount = 5

    private val tabOpenSource = AtomicReference<TabLayout.Tab>()
    private val tabPrivacy = AtomicReference<TabLayout.Tab>()
    private val tabAuthData = AtomicReference<TabLayout.Tab>()

    private lateinit var binding: ActivitySetupViewerBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        binding = ActivitySetupViewerBinding.inflate(layoutInflater)
        setContentView(binding.root)
        ViewCompat.setOnApplyWindowInsetsListener(binding.root) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }
        vpAdapter = ViewPagerAdapter(supportFragmentManager, lifecycle)

        // Add base fragments and tabs
        vpAdapter.addFragment(pageLanguage, PAGE_LABEL_LANGUAGE)
        vpAdapter.addFragment(pageConsent, PAGE_LABEL_CONSENT)

        binding.tablayout.addTab(binding.tablayout.newTab().setText(PAGE_LABEL_LANGUAGE))
        binding.tablayout.addTab(binding.tablayout.newTab().setText(PAGE_LABEL_CONSENT))

        binding.viewpager.setAdapter(vpAdapter)

        // Create page indicators and select current page
        repeat(pageCount) {
            val dotPage = ImageView(this).apply {
                scaleType = ImageView.ScaleType.FIT_CENTER
                backgroundTintList = attrColorStateList(R.attr.colorSurfaceContainerHigh)
                setBackgroundResource(R.drawable.page_dot)
            }
            val dp8 = 8.dpToPx
            val dp2 = 2.dpToPx
            val lp = RelativeLayout.LayoutParams(dp8, dp8).apply {
                setMargins(dp2, 0, dp2, 0)
            }
            binding.pageDotContainer.addView(dotPage, lp)
            pageDots.add(dotPage)
        }
        pageDots[0].backgroundTintList = attrColorStateList(R.attr.colorPrimary)

        // Setup page action listeners
        pageLanguage.onLanguageChangedListener = ::languageChanged
        pageConsent.setOnConsentChangedListener(::consentChanged)
        pagePrivacy.setOnAuthTypeChangedListener { authType ->
            selectedAuthType = authType
            pageAuthentication.showCredentialsSetup(authType)
        }

        // Setup viewer action listeners
        binding.tablayout.addOnTabSelectedListener(tabSelected())
        binding.viewpager.registerOnPageChangeCallback(changeTab())
        binding.btnNext.setOnClickListener(::progressToNextPage)
    }

    private fun progressToNextPage(view: View) {
        // If the current page is the last one (the authentication configuration page) then store
        // the authentication and proceed to the get started page to finish the setup process
        if (vpAdapter.isCurrentPage(PAGE_LABEL_AUTHENTICATION, binding.tablayout.selectedTabPosition)) {
            lifecycleScope.launch(Dispatchers.IO) {
                when (selectedAuthType) {
                    AuthTypes.Password -> storePassword()
                    AuthTypes.Pin -> storePIN()
                    AuthTypes.None -> storeWithoutAuthentication()
                }
                updateSetting(DataStoreKeys.APP_SETUP_FINISHED, true)
                runOnUiThread {
                    startActivity(Intent(this@SetupViewer, SetupGetStartedView::class.java).apply {
                        setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                    })
                    finish()
                }
            }
            return
        }

        // Ensure the next page is in a valid range
        val nextPagePosition = binding.tablayout.selectedTabPosition + 1
        if (nextPagePosition >= vpAdapter.itemCount) {
            return
        }

        // Move to the next page
        binding.tablayout.selectTab(binding.tablayout.getTabAt(nextPagePosition))
    }

    private suspend fun storeWithoutAuthentication() {
        updateSetting(DataStoreKeys.AUTHENTICATION_TYPE, "none")
    }

    private suspend fun storePIN() {
        if (pageAuthentication.pin.isEmpty()) {
            Toast.makeText(this@SetupViewer, getResources().getString(R.string.pin_too_short), Toast.LENGTH_SHORT).show()
            return
        }
        val secretKey = Crypt.generateSecretKey()
        val pinCipher = Crypt.encryptStringBlowfish(pageAuthentication.pin, secretKey)
        updateSetting(DataStoreKeys.AUTHENTICATION_TYPE, "pin")
        updateSetting(DataStoreKeys.AUTHENTICATION_HASH, pinCipher)
        updateSetting(DataStoreKeys.AUTHENTICATION_SALT, Base64.encodeToString(secretKey, Base64.NO_WRAP))
        updateSetting(DataStoreKeys.AUTHENTICATION_USE_BIOMETRICS, pagePrivacy.useBiometrics)
    }

    private suspend fun storePassword() {
        if (pageAuthentication.password.isEmpty()) {
            Toast.makeText(this@SetupViewer, getResources().getString(R.string.password_too_short), Toast.LENGTH_SHORT).show()
            return
        }
        val salt = Crypt.generateSalt()
        val pwHash = Crypt.encryptString(pageAuthentication.password, salt)
        updateSetting(DataStoreKeys.AUTHENTICATION_TYPE, "password")
        updateSetting(DataStoreKeys.AUTHENTICATION_HASH, pwHash)
        updateSetting(DataStoreKeys.AUTHENTICATION_SALT, Base64.encodeToString(salt, Base64.NO_WRAP))
        updateSetting(DataStoreKeys.AUTHENTICATION_USE_BIOMETRICS, pagePrivacy.useBiometrics)
    }

    private fun changeTab(): OnPageChangeCallback = object : OnPageChangeCallback() {
        override fun onPageSelected(position: Int) {
            binding.tablayout.selectTab(binding.tablayout.getTabAt(position))
        }
    }

    private fun tabSelected(): OnTabSelectedListener = object : OnTabSelectedListener {
        override fun onTabSelected(tab: TabLayout.Tab) {
            binding.viewpager.currentItem = tab.position
            pageDots[tab.position].backgroundTintList = attrColorStateList(R.attr.colorPrimary)

            val isLastPage = binding.tablayout.selectedTabPosition + 1 == pageCount
            if (isLastPage) {
                binding.btnNext.setText(R.string.setup_next_finish)
                binding.btnNext.setIcon(null)
            } else {
                binding.btnNext.setText(R.string.setup_next)
                binding.btnNext.setIcon(resolveDrawable(R.drawable.rounded_chevron_right_24))
            }
        }

        override fun onTabUnselected(tab: TabLayout.Tab) {
            pageDots[tab.position].backgroundTintList = attrColorStateList(R.attr.colorSurfaceContainerHigh)
        }

        override fun onTabReselected(tab: TabLayout.Tab) { }
    }

    private fun consentChanged(currentConsent: Boolean) {
        if (currentConsent) {
            vpAdapter.addFragment(pageOpenSource, PAGE_LABEL_OPEN_SOURCE)
            vpAdapter.addFragment(pagePrivacy, PAGE_LABEL_PRIVACY)
            vpAdapter.addFragment(pageAuthentication, PAGE_LABEL_AUTHENTICATION)

            tabOpenSource.set(binding.tablayout.newTab().setText(PAGE_LABEL_OPEN_SOURCE))
            tabPrivacy.set(binding.tablayout.newTab().setText(PAGE_LABEL_PRIVACY))
            tabAuthData.set(binding.tablayout.newTab().setText(PAGE_LABEL_AUTHENTICATION))

            binding.tablayout.addTab(tabOpenSource.get())
            binding.tablayout.addTab(tabPrivacy.get())
            binding.tablayout.addTab(tabAuthData.get())
        } else {
            vpAdapter.removeFragment(pageOpenSource, PAGE_LABEL_OPEN_SOURCE)
            vpAdapter.removeFragment(pagePrivacy, PAGE_LABEL_PRIVACY)
            vpAdapter.removeFragment(pageAuthentication, PAGE_LABEL_AUTHENTICATION)

            binding.tablayout.removeTab(tabOpenSource.get())
            binding.tablayout.removeTab(tabPrivacy.get())
            binding.tablayout.removeTab(tabAuthData.get())

        }
        consented.set(currentConsent)
    }

    private fun languageChanged() = runOnUiThread {
        binding.btnNext.text = resources.getString(R.string.setup_next)
        (vpAdapter.getFragment(PAGE_LABEL_LANGUAGE) as SetupLanguageView).onLanguageUpdated()
        (vpAdapter.getFragment(PAGE_LABEL_CONSENT) as SetupConsentView).onLanguageUpdated()
        if (consented.get()) {
            (vpAdapter.getFragment(PAGE_LABEL_OPEN_SOURCE) as SetupOpenSourceView).onLanguageUpdated()
            (vpAdapter.getFragment(PAGE_LABEL_PRIVACY) as SetupPrivacyView).onLanguageUpdated()
            (vpAdapter.getFragment(PAGE_LABEL_AUTHENTICATION) as SetupAuthDataView).onLanguageUpdated()
        }
    }
}